feat(amazonq): add commands to command palette

ivikash opened the pull request:
## Problem
Add commands in Command Palette for Amazon Q to improve feature discoverability

## Solution
Commands Added
- Amazon Q: Open Chat Panel
- Amazon Q: Open Code Reference Log
- Amazon Q: Select Customizations
- Amazon Q: Toggle Auto-Scans
- Amazon Q: Toggle Auto-Suggestions
- Amazon Q: Try inline suggestion example
- Amazon Q: Sign Out

TODO:
- [X] ~~Discuss `FromCommandPalette` pattern. -- Discussed offline, this is the right approach. Going forward use the same prefix and add `.commandPalette` as suffix.~~
- [X] Discuss `telemetry`/`metrics`

## License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

ivikash commented:
## High Level Concerns
- Invalid `source` - example - <pre>2024-05-10 09:13:39.725 [error] Commands/Telemetry: "aws.amazonq.signout" executed with invalid "source" type: "[]"<br />2024-05-10 09:13:39.725 [debug] command: running "aws.amazonq.signout" with arguments: [ <1 empty item>, 'sourceImproperlySet' ]<br />2024-05-10 09:13:39.726 [debug] command: running "aws.amazonq.focusChat" with arguments: [ 'placeholder**', 'sourceImproperlySet' ]</pre>
- Duplicating commands in `package.json` with suffix `FromCommandPalette`.


## Metric Details

| **Command / Label**                       | **Metric**                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                            | **Completed** | **Error**                                                                                                                                                                                                                                                                                                                                                                                                     |
|-------------------------------------------|-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|---------------|---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Amazon Q: Sign out                        | <pre>2024-05-10 09:13:39.725 [error] Commands/Telemetry: "aws.amazonq.signout" executed with invalid "source" type: "[]"<br>2024-05-10 09:13:39.725 [debug] command: running "aws.amazonq.signout" with arguments: [ <1 empty item>, 'sourceImproperlySet' ]<br>2024-05-10 09:13:39.726 [debug] command: running "aws.amazonq.focusChat" with arguments: [ 'placeholder**', 'sourceImproperlySet' ]<br>2024-05-10 09:13:39.748 [debug] telemetry: vscode_executeCommand {<br>  Metadata: {<br>    parentMetric: 'vscode_executeCommand',<br>    command: 'aws.amazonq.focusChat',<br>    source: 'sourceImproperlySet',<br>    duration: '22',<br>    result: 'Succeeded',<br>    awsAccount: 'not-set',<br>    awsRegion: 'us-east-1'<br>  },<br>  Value: 1,<br>  Unit: 'None',<br>  Passive: true<br>}<br>2024-05-10 09:13:39.748 [debug] telemetry: vscode_executeCommand {<br>  Metadata: {<br>    command: 'aws.amazonq.signout',<br>    source: 'sourceImproperlySet',<br>    duration: '24',<br>    result: 'Succeeded',<br>    awsAccount: 'not-set',<br>    awsRegion: 'us-east-1'<br>  },<br>  Value: 1,<br>  Unit: 'None',<br>  Passive: true<br>}<br>2024-05-10 09:13:39.754 [debug] telemetry: auth_signInPageOpened {<br>  Metadata: {<br>    result: 'Succeeded',<br>    awsAccount: 'not-set',<br>    awsRegion: 'us-east-1'<br>  },<br>  Value: 1,<br>  Unit: 'None',<br>  Passive: true<br>}<br>2024-05-10 09:13:39.859 [debug] Registered event handler for: onActiveConnectionModified<br>2024-05-10 09:13:39.859 [debug] Webview called command "refreshAuthState" with args: []<br>2024-05-10 09:13:39.863 [debug] Webview called command "getAuthState" with args: []<br>2024-05-10 09:13:39.869 [debug] Webview called command "getDefaultStartUrl" with args: []<br>2024-05-10 09:13:39.872 [debug] Webview called command "getRegions" with args: []<br>2024-05-10 09:13:39.872 [debug] Webview called command "fetchConnections" with args: []<br>2024-05-10 09:13:39.872 [debug] Webview called command "resetStoredMetricMetadata" with args: []<br>2024-05-10 09:13:39.897 [debug] Webview called command "listConnections" with args: []<br>2024-05-10 09:18:21.209 [info] telemetry: sent batch (size=6)<br></pre>                                                                    | ❌             | <pre>2024-05-10 09:13:39.725 [error] Commands/Telemetry: "aws.amazonq.signout" executed with invalid "source" type: "[]"<br />2024-05-10 09:13:39.725 [debug] command: running "aws.amazonq.signout" with arguments: [ <1 empty item>, 'sourceImproperlySet' ]<br />2024-05-10 09:13:39.726 [debug] command: running "aws.amazonq.focusChat" with arguments: [ 'placeholder**', 'sourceImproperlySet' ]</pre> |
| Amazon Q: Open Code Reference Log         | <pre>2024-05-10 09:36:01.983 [debug] command: running "aws.amazonq.openReferencePanelFromCommandPalette" with arguments: []<br />2024-05-10 09:36:01.984 [debug] command: running "aws.amazonq.openReferencePanel" with arguments: [ 'placeholder**', 'CommandPalette' ]<br />2024-05-10 09:36:01.997 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    command: 'aws.amazonq.openReferencePanel',<br />    source: 'CommandPalette',<br />    duration: '14',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:36:01.997 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.openReferencePanelFromCommandPalette',<br />    duration: '14',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br /></pre>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         | ✅             |                                                                                                                                                                                                                                                                                                                                                                                                               |
| Amazon Q: Select Customization            | <pre>2024-05-10 09:37:04.247 [error] Commands/Telemetry: "aws.amazonq.selectCustomization" executed with invalid "source" type: "[]"<br />2024-05-10 09:37:04.247 [debug] command: running "aws.amazonq.selectCustomization" with arguments: [ <1 empty item>, 'sourceImproperlySet' ]<br />2024-05-10 09:37:04.247 [debug] telemetry: ui_click {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    elementId: 'cw_selectCustomization_Cta',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-10 09:37:04.247 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.selectCustomization',<br />    source: 'sourceImproperlySet',<br />    duration: '1',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:37:04.251 [debug] icons: using override for "vscode-help"<br />2024-05-10 09:37:04.263 [debug] SSO token cache: loaded key: 4ceb1ad5-5395-445d-828f-94c1fe90478d<br />2024-05-10 09:37:04.263 [info] current client registration id=_ULgRtooSoe1lvaNHR8mhHVzLWVhc3QtMQ, <br />    expires at Thu Aug 08 2024 09:35:23 GMT-0700 (Pacific Daylight Time), <br />    key = 4ceb1ad5-5395-445d-828f-94c1fe90478d<br />2024-05-10 09:37:05.926 [debug] amazonq: listAvailableCustomizations API request:<br />    RequestID: c4bb17fe-e95b-4de6-8743-70bc095f16a5,<br />    Customizations:<br />        00: Amazon-Internal-V1<br />        01: NGDE-Custom-Finetuning-7K-V2<br />        02: NGDE-Custom-V13<br /></pre>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | ❌             | <pre>2024-05-10 09:37:04.247 [error] Commands/Telemetry: "aws.amazonq.selectCustomization" executed with invalid "source" type: "[]"<br />2024-05-10 09:37:04.247 [debug] command: running "aws.amazonq.selectCustomization" with arguments: [ <1 empty item>, 'sourceImproperlySet' ]</pre>                                                                                                                  |
| Amazon Q: Try inline suggestion examples  | <pre>2024-05-10 09:37:44.203 [debug] command: running "CodeWhispererCommandBackend.showGettingStartedPage" (id: aws.amazonq.gettingStarted) with arguments: []<br />2024-05-10 09:37:44.203 [debug] telemetry: ui_click {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    elementId: 'codewhisperer_Learn_PageOpen',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:37:44.204 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.gettingStarted',<br />    duration: '4',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:37:44.325 [debug] Webview called command "getOSType" with args: []<br /></pre>`                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          | ✅             |                                                                                                                                                                                                                                                                                                                                                                                                               |
| Amazon Q: Open Chat Panel                 | <pre>2024-05-10 09:38:06.945 [debug] command: running "aws.amazonq.openChatPanelFromCommandPalette" with arguments: []<br />2024-05-10 09:38:06.945 [debug] command: running "aws.amazonq.openChatPanel" with arguments: [ 'placeholder**', 'CommandPalette' ]<br />2024-05-10 09:38:06.945 [error] Commands/Telemetry: "aws.amazonq.focusChat" executed with invalid "source" type: "[]"<br />2024-05-10 09:38:06.945 [debug] command: running "aws.amazonq.focusChat" with arguments: [ <1 empty item>, 'sourceImproperlySet' ]<br />2024-05-10 09:38:06.947 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    command: 'aws.amazonq.focusChat',<br />    source: 'sourceImproperlySet',<br />    duration: '1',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:38:06.947 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    command: 'aws.amazonq.openChatPanel',<br />    source: 'CommandPalette',<br />    duration: '2',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:38:06.947 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.openChatPanelFromCommandPalette',<br />    duration: '2',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:38:07.000 [debug] telemetry: amazonq_enterFocusChat {<br />  Metadata: {<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:38:08.535 [debug] telemetry: amazonq_exitFocusChat {<br />  Metadata: {<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br /></pre> | ❌             | <pre>2024-05-10 09:38:06.945 [error] Commands/Telemetry: "aws.amazonq.focusChat" executed with invalid "source" type: "[]"<br />2024-05-10 09:38:06.945 [debug] command: running "aws.amazonq.focusChat" with arguments: [ <1 empty item>, 'sourceImproperlySet' ]</pre>                                                                                                                                      |
| Amazon Q: Resume Auto-Suggestions         | <pre>2024-05-10 09:40:13.946 [debug] command: running "aws.amazonq.resumeAutoSuggestions" with arguments: []<br />2024-05-10 09:40:13.946 [debug] telemetry: collapsing 1 "aws.amazonq.toggleCodeSuggestion" metrics. key={ '1': 'source' }<br />2024-05-10 09:40:13.946 [debug] command: running "aws.amazonq.toggleCodeSuggestion" with arguments: [ 'placeholder**', 'CommandPalette' ]<br />2024-05-10 09:40:13.952 [debug] telemetry: aws_modifySetting {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    settingId: 'codewhisperer_autoSuggestionActivation',<br />    settingState: 'Activated',<br />    duration: '6',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-10 09:40:13.952 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.resumeAutoSuggestions',<br />    duration: '8',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br /></pre>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                | ✅             |                                                                                                                                                                                                                                                                                                                                                                                                               |
| Amazon Q: Pause Auto-Suggestions          | <pre>2024-05-10 09:39:50.151 [debug] command: running "aws.amazonq.pauseAutoSuggestions" with arguments: []<br />2024-05-10 09:39:50.151 [debug] command: running "aws.amazonq.toggleCodeSuggestion" with arguments: [ 'placeholder**', 'CommandPalette' ]<br />2024-05-10 09:39:50.159 [debug] telemetry: aws_modifySetting {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    settingId: 'codewhisperer_autoSuggestionActivation',<br />    settingState: 'Deactivated',<br />    duration: '9',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-10 09:39:50.159 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    command: 'aws.amazonq.toggleCodeSuggestion',<br />    source: 'CommandPalette',<br />    duration: '9',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:39:50.159 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.pauseAutoSuggestions',<br />    duration: '10',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}</pre>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                              | ✅             |                                                                                                                                                                                                                                                                                                                                                                                                               |
| Amazon Q: Resume Auto-Scans               | <pre>2024-05-10 09:39:07.165 [debug] command: running "aws.amazonq.resumeAutoScans" with arguments: []<br />2024-05-10 09:39:07.165 [debug] command: running "aws.codeWhisperer.toggleCodeScan" with arguments: [ 'placeholder**', 'CommandPalette' ]<br />2024-05-10 09:39:07.177 [debug] telemetry: aws_modifySetting {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    settingId: 'codewhisperer_autoScansActivation',<br />    settingState: 'Activated',<br />    duration: '12',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-10 09:39:07.177 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    command: 'aws.codeWhisperer.toggleCodeScan',<br />    source: 'CommandPalette',<br />    duration: '12',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-10 09:39:07.177 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.resumeAutoScans',<br />    duration: '14',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}</pre>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                             | ✅             |                                                                                                                                                                                                                                                                                                                                                                                                               |
| Amazon Q: Pause Auto-Scan                 | <pre>2024-05-10 09:39:29.603 [debug] command: running "aws.amazonq.pauseAutoScans" with arguments: []<br />2024-05-10 09:39:29.603 [debug] telemetry: collapsing 1 "aws.codeWhisperer.toggleCodeScan" metrics. key={ '1': 'source' }<br />2024-05-10 09:39:29.603 [debug] command: running "aws.codeWhisperer.toggleCodeScan" with arguments: [ 'placeholder**', 'CommandPalette' ]<br />2024-05-10 09:39:29.608 [debug] telemetry: aws_modifySetting {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    settingId: 'codewhisperer_autoScansActivation',<br />    settingState: 'Deactivated',<br />    duration: '6',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-10 09:39:29.609 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.pauseAutoScans',<br />    duration: '7',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br /></pre>                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 | ✅             |                                                                                                                                                                                                                                                                                                                                                                                                               |

justinmk3 commented on packages/amazonq/package.json:
```diff
@@ -614,6 +607,45 @@
+                "enablement": "aws.codewhisperer.isSuggestionsEnabled"
+            },
+            {
+                "command": "aws.amazonq.resumeAutoScans",
+                "title": "%AWS.amazonq.resumeAutoScans%",
+                "category": "%AWS.amazonq.title%",
+                "enablement": "!aws.codewhisperer.isScanEnabled"
+            },
+            {
+                "command": "aws.amazonq.pauseAutoScans",
```
can these be one "toggle" command instead of 2 separate commands? that allows users to define a single keybinding

justinmk3 commented on packages/amazonq/package.json:
```diff
@@ -614,6 +607,45 @@
+                "category": "%AWS.amazonq.title%"
+            },
+            {
+                "command": "aws.amazonq.resumeAutoSuggestions",
+                "title": "%AWS.amazonq.resumeAutoSuggestions%",
+                "category": "%AWS.amazonq.title%",
+                "enablement": "!aws.codewhisperer.isSuggestionsEnabled"
+            },
+            {
+                "command": "aws.amazonq.pauseAutoSuggestions",
```
same question for these 2

justinmk3 commented on packages/amazonq/package.json:
```diff
@@ -614,6 +607,45 @@
+                "category": "%AWS.amazonq.title%"
+            },
+            {
+                "command": "aws.amazonq.gettingStarted",
```
why does this need to be the command palette? vscode already exposes all walkthroughs via `Welcome: Open Walkthrough...`

justinmk3 commented on packages/core/src/amazonq/activation.ts:
```diff
@@ -50,6 +52,12 @@ export async function activate(context: ExtensionContext) {
         void vscode.env.openExternal(vscode.Uri.parse(amazonQHelpUrl))
     })
 
+    const isSuggestionsEnabled = CodeSuggestionsState.instance.isSuggestionsEnabled()
+    await vscode.commands.executeCommand('setContext', 'aws.codewhisperer.isSuggestionsEnabled', isSuggestionsEnabled)
+
+    const isScansEnabled = CodeScansState.instance.isScansEnabled()
+    await vscode.commands.executeCommand('setContext', 'aws.codewhisperer.isScanEnabled', isScansEnabled)
```
wouldn't need these if we had "toggle" command instead

justinmk3 commented on packages/core/src/codewhisperer/commands/basicCommands.ts:
```diff
@@ -84,13 +90,15 @@ export const toggleCodeScans = Commands.declare(
             })
 
             const isScansEnabled = await scansState.toggleScans()
+            await vscode.commands.executeCommand('setContext', 'aws.codewhisperer.isScanEnabled', isScansEnabled)
```
wouldn't need this either, if we have the toggle command

justinmk3 commented on packages/core/src/codewhisperer/commands/types.ts:
```diff
@@ -16,6 +16,8 @@ export const amazonQChatSource = 'amazonQChat'
 /** Indicates a CodeWhisperer command was executed as a result of selecting an ellipses menu item */
 export const cwEllipsesMenu = 'ellipsesMenu'
+/** Indicates a CodeWhisperer command was executed from the command palette */
+export const commandPalette = 'CommandPalette'
```
please, always use the existing code as a hint about conventions. none of the other ids above use PascalCase
```suggestion
export const commandPalette = 'commandPalette'
```

ivikash replied:
Certainly. To clarify, I initially preferred to use camelCase in this context. However, upon reviewing the existing code, I noticed and believed that the `CommandPalette` convention is already implemented as seen in here - https://github.com/aws/aws-toolkit-vscode/blob/fade41d86565087ac2465c3ade219f552c745aeb/packages/core/src/codecatalyst/commands.ts#L295

I will update it in the next revision.

justinmk3 replied:
> the `CommandPalette` convention is already implemented as seen in here

Great, thank you for noting that. In that case, yes we should keep the existing name, but please update that code to use the constant you defined here.

ivikash replied:
I saw another place relying on `commandPalette` so I am erring on the side of using lowecase and updating `codecatalyst/commands.ts` to use the same with smaller case.

I am making sure all instances use the same variable.

ivikash commented:
## Demo

https://github.com/aws/aws-toolkit-vscode/assets/371007/eb4d9dce-b43f-4de4-add6-368bd4baaed8

## Metrics

| Command | Telemetry |
|--|--|
| Amazon Q: Open Code Reference Log | <pre>2024-05-26 19:15:09.358 [debug] command: running "aws.amazonq.openReferencePanel.commandPalette" with arguments: []<br />2024-05-26 19:15:09.359 [debug] command: running "aws.amazonq.openReferencePanel" with arguments: [ 'placeholder**', 'commandPalette' ]<br />2024-05-26 19:15:09.381 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    command: 'aws.amazonq.openReferencePanel',<br />    source: 'commandPalette',<br />    duration: '22',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:15:09.381 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.openReferencePanel.commandPalette',<br />    duration: '24',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br /></pre>  |
| Amazon Q: Try inline suggestion example | <pre>2024-05-26 19:15:53.849 [debug] command: running "CodeWhispererCommandBackend.showGettingStartedPage" (id: aws.amazonq.gettingStarted) with arguments: []<br />2024-05-26 19:15:53.849 [debug] telemetry: ui_click {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    elementId: 'codewhisperer_Learn_PageOpen',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:15:53.849 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.gettingStarted',<br />    duration: '3',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:15:53.967 [debug] Webview called command "getOSType" with args: []<br /></pre>  |
| Amazon Q: Open Chat Panel |  <pre>2024-05-26 19:16:17.774 [debug] command: running "aws.amazonq.openChatPanel.commandPalette" with arguments: []<br />2024-05-26 19:16:17.774 [debug] command: running "aws.amazonq.focusChat" with arguments: [ 'placeholder**', 'commandPalette' ]<br />2024-05-26 19:16:17.795 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    command: 'aws.amazonq.focusChat',<br />    source: 'commandPalette',<br />    duration: '22',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:16:17.796 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.openChatPanel.commandPalette',<br />    duration: '23',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:16:17.799 [debug] SSO token cache: loaded key: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:16:17.799 [info] current client registration id=BKZm_Lf95sS6wrGZnexbbHVzLWVhc3QtMQ, <br />    expires at Tue Aug 20 2024 15:10:29 GMT-0700 (Pacific Daylight Time), <br />    key = 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:16:17.799 [info] auth: Connection is valid: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:16:17.799 [info] auth: Updating connection state of 06f56967-f6e1-47f0-9ff2-5e95a9b93d10 to valid<br />2024-05-26 19:16:18.586 [debug] telemetry: amazonq_exitFocusChat {<br />  Metadata: {<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br /></pre> |
| Amazon Q: Select Customizations | <pre>2024-05-26 19:17:09.423 [debug] telemetry: collapsing 1 "aws.amazonq.selectCustomization.commandPalette" metrics. key={}<br />2024-05-26 19:17:09.424 [debug] command: running "aws.amazonq.selectCustomization.commandPalette" with arguments: []<br />2024-05-26 19:17:09.424 [debug] telemetry: collapsing 1 "aws.amazonq.selectCustomization" metrics. key={ '1': 'source' }<br />2024-05-26 19:17:09.424 [debug] command: running "aws.amazonq.selectCustomization" with arguments: [ 'placeholder**', 'commandPalette' ]<br />2024-05-26 19:17:09.424 [debug] telemetry: ui_click {<br />  Metadata: {<br />    elementId: 'cw_selectCustomization_Cta',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-26 19:17:09.449 [debug] SSO token cache: loaded key: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:17:09.449 [info] current client registration id=BKZm_Lf95sS6wrGZnexbbHVzLWVhc3QtMQ, <br />    expires at Tue Aug 20 2024 15:10:29 GMT-0700 (Pacific Daylight Time), <br />    key = 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:17:11.061 [debug] amazonq: listAvailableCustomizations API request:<br />    RequestID: a513a335-6ea1-4f9f-8e2c-17a0d44e324a,<br />    Customizations:<br />        00: Amazon-Internal-V1<br />        01: NGDE-Custom-Finetuning-7K-V2<br />        02: NGDE-Custom-V13<br /></pre>  |
| Amazon Q: Toggle Auto-Scans | <pre>2024-05-26 19:17:51.082 [debug] telemetry: collapsing 2 "aws.codeWhisperer.toggleCodeScan.commandPalette" metrics. key={}<br />2024-05-26 19:17:51.083 [debug] command: running "aws.codeWhisperer.toggleCodeScan.commandPalette" with arguments: []<br />2024-05-26 19:17:51.083 [debug] telemetry: collapsing 2 "aws.codeWhisperer.toggleCodeScan" metrics. key={ '1': 'source' }<br />2024-05-26 19:17:51.083 [debug] command: running "aws.codeWhisperer.toggleCodeScan" with arguments: [ 'placeholder**', 'commandPalette' ]<br />2024-05-26 19:17:51.095 [debug] telemetry: aws_modifySetting {<br />  Metadata: {<br />    settingId: 'codewhisperer_autoScansActivation',<br />    settingState: 'Activated',<br />    duration: '14',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-26 19:17:53.523 [debug] telemetry: collapsing 1 "aws.amazonq.listCommands" metrics. key={}<br />2024-05-26 19:17:53.523 [debug] command: running "aws.amazonq.listCommands" with arguments: []<br />2024-05-26 19:17:53.523 [debug] telemetry: ui_click {<br />  Metadata: {<br />    elementId: 'cw_statusBarMenu',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-26 19:17:53.524 [debug] telemetry: collapsing 1 "aws.amazonq.refreshAnnotation" metrics. key={}<br />2024-05-26 19:17:53.524 [debug] command: running "aws.amazonq.refreshAnnotation" with arguments: [ true ]<br />2024-05-26 19:17:57.545 [debug] telemetry: collapsing 3 "aws.codeWhisperer.toggleCodeScan.commandPalette" metrics. key={}<br />2024-05-26 19:17:57.546 [debug] command: running "aws.codeWhisperer.toggleCodeScan.commandPalette" with arguments: []<br />2024-05-26 19:17:57.546 [debug] telemetry: collapsing 3 "aws.codeWhisperer.toggleCodeScan" metrics. key={ '1': 'source' }<br />2024-05-26 19:17:57.547 [debug] command: running "aws.codeWhisperer.toggleCodeScan" with arguments: [ 'placeholder**', 'commandPalette' ]<br />2024-05-26 19:17:57.557 [debug] telemetry: aws_modifySetting {<br />  Metadata: {<br />    settingId: 'codewhisperer_autoScansActivation',<br />    settingState: 'Deactivated',<br />    duration: '12',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-26 19:17:58.616 [debug] telemetry: collapsing 2 "aws.amazonq.listCommands" metrics. key={}<br />2024-05-26 19:17:58.616 [debug] command: running "aws.amazonq.listCommands" with arguments: []<br />2024-05-26 19:17:58.617 [debug] telemetry: ui_click {<br />  Metadata: {<br />    elementId: 'cw_statusBarMenu',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-26 19:17:58.617 [debug] telemetry: collapsing 2 "aws.amazonq.refreshAnnotation" metrics. key={}<br />2024-05-26 19:17:58.617 [debug] command: running "aws.amazonq.refreshAnnotation" with arguments: [ true ]<br />2024-05-26 19:17:59.887 [debug] telemetry: amazonq_enterFocusChat {<br />  Metadata: {<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:18:00.425 [debug] telemetry: amazonq_exitFocusChat {<br />  Metadata: {<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br /></pre>  |
| Amazon Q: Toggle Auto-Suggestions | <pre>2024-05-26 19:18:25.388 [debug] command: running "aws.amazonq.toggleCodeSuggestion.commandPalette" with arguments: []<br />2024-05-26 19:18:25.388 [debug] command: running "aws.amazonq.toggleCodeSuggestion" with arguments: [ 'placeholder**', 'commandPalette' ]<br />2024-05-26 19:18:25.391 [debug] telemetry: aws_modifySetting {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    settingId: 'codewhisperer_autoSuggestionActivation',<br />    settingState: 'Deactivated',<br />    duration: '3',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br />2024-05-26 19:18:25.391 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    command: 'aws.amazonq.toggleCodeSuggestion',<br />    source: 'commandPalette',<br />    duration: '3',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:18:25.391 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.toggleCodeSuggestion.commandPalette',<br />    duration: '5',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:18:26.932 [debug] telemetry: collapsing 1 "aws.amazonq.toggleCodeSuggestion.commandPalette" metrics. key={}<br />2024-05-26 19:18:26.932 [debug] command: running "aws.amazonq.toggleCodeSuggestion.commandPalette" with arguments: []<br />2024-05-26 19:18:26.932 [debug] telemetry: collapsing 1 "aws.amazonq.toggleCodeSuggestion" metrics. key={ '1': 'source' }<br />2024-05-26 19:18:26.932 [debug] command: running "aws.amazonq.toggleCodeSuggestion" with arguments: [ 'placeholder**', 'commandPalette' ]<br />2024-05-26 19:18:26.935 [debug] telemetry: aws_modifySetting {<br />  Metadata: {<br />    settingId: 'codewhisperer_autoSuggestionActivation',<br />    settingState: 'Activated',<br />    duration: '3',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'us-east-1'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: false<br />}<br /></pre>  |
| Amazon Q: Sign Out | <pre>2024-05-26 19:18:48.772 [debug] command: running "aws.amazonq.signout.commandPalette" with arguments: []<br />2024-05-26 19:18:48.773 [debug] command: running "aws.amazonq.signout" with arguments: [ 'placeholder**', 'commandPalette' ]<br />2024-05-26 19:18:48.773 [info] auth: logout<br />2024-05-26 19:18:48.779 [info] auth: Invalidating connection: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:48.796 [debug] SSO token cache: loaded key: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:48.796 [info] current client registration id=BKZm_Lf95sS6wrGZnexbbHVzLWVhc3QtMQ, <br />    expires at Tue Aug 20 2024 15:10:29 GMT-0700 (Pacific Daylight Time), <br />    key = 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.118 [warning] credentials (sso): invalidating stored token: Session token not found or invalid<br />2024-05-26 19:18:49.118 [info] SsoAccessTokenProvider invalidate token and registration<br />2024-05-26 19:18:49.123 [debug] SSO token cache: deleted (reason: SsoAccessTokenProvider.invalidate()) key: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.127 [debug] auth: SSO registration cache: deleted (reason: SsoAccessTokenProvider.invalidate()) key: {"startUrl":"https://amzn.awsapps.com/start","region":"us-east-1","scopes":["codewhisperer:completions","codewhisperer:analysis","codewhisperer:conversations","codewhisperer:transformations","codewhisperer:taskassist"]}<br />2024-05-26 19:18:49.127 [warning] auth: failed to logout of connection "06f56967-f6e1-47f0-9ff2-5e95a9b93d10": UnauthorizedException: Session token not found or invalid<br />2024-05-26 19:18:49.127 [info] SsoAccessTokenProvider invalidate token and registration<br />2024-05-26 19:18:49.133 [debug] SSO token cache: file not found key: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.133 [debug] auth: SSO registration cache: file not found key: {"startUrl":"https://amzn.awsapps.com/start","region":"us-east-1","scopes":["codewhisperer:completions","codewhisperer:analysis","codewhisperer:conversations","codewhisperer:transformations","codewhisperer:taskassist"]}<br />2024-05-26 19:18:49.133 [info] auth: Updating connection state of 06f56967-f6e1-47f0-9ff2-5e95a9b93d10 to invalid<br />2024-05-26 19:18:49.146 [info] codewhisperer: active connection changed<br />2024-05-26 19:18:49.146 [info] secondaryAuth: connectionId=06f56967-f6e1-47f0-9ff2-5e95a9b93d10, connectionStatus=invalid<br />2024-05-26 19:18:49.146 [info] codewhisperer: connection states<br />    connection isValid=false,<br />    connection isValidCodewhispererCoreConnection=true,<br />    connection isExpired=true,<br />    secondaryAuth isExpired=true,<br />    connection isUndefined=false<br />2024-05-26 19:18:49.146 [info] codewhisperer: connection changed to invalid: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.146 [info] toolkitApi: set connection 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.149 [info] secondaryAuth: connectionId=06f56967-f6e1-47f0-9ff2-5e95a9b93d10, connectionStatus=undefined<br />2024-05-26 19:18:49.150 [info] codewhisperer: connection states<br />    connection isValid=true,<br />    connection isValidCodewhispererCoreConnection=true,<br />    connection isExpired=false,<br />    secondaryAuth isExpired=false,<br />    connection isUndefined=false<br />2024-05-26 19:18:49.150 [debug] amazonq: Fetching feature configs<br />2024-05-26 19:18:49.154 [info] toolkitApi: delete connection 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.160 [debug] SSO token cache: read failed (file not found) key: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.160 [info] current client registration id=undefined, <br />    expires at undefined, <br />    key = 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.160 [info] auth: Handling invalid credentials of connection: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.160 [info] auth: Updating connection state of 06f56967-f6e1-47f0-9ff2-5e95a9b93d10 to invalid<br />2024-05-26 19:18:49.160 [error] CodeWhisperer: Error when fetching feature configs Error: Profile does not exist: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10 Profile does not exist: 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.160 [debug] CodeWhisperer: Current feature configs: {NewEOS: TREATMENT, 17B: DEFAULT}<br />2024-05-26 19:18:49.167 [info] codewhisperer: active connection changed<br />2024-05-26 19:18:49.168 [info] codewhisperer: connection states<br />    connection isValid=false,<br />    connection isValidCodewhispererCoreConnection=false,<br />    connection isExpired=false,<br />    secondaryAuth isExpired=false,<br />    connection isUndefined=true<br />2024-05-26 19:18:49.168 [info] codewhisperer: active connection changed<br />2024-05-26 19:18:49.168 [debug] telemetry: collapsing 1 "aws.amazonq.focusChat" metrics. key={ '1': 'source' }<br />2024-05-26 19:18:49.168 [debug] command: running "aws.amazonq.focusChat" with arguments: [ 'placeholder**', 'commandPalette' ]<br />2024-05-26 19:18:49.181 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    parentMetric: 'vscode_executeCommand',<br />    command: 'aws.amazonq.signout',<br />    source: 'commandPalette',<br />    duration: '410',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'not-set'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:18:49.181 [debug] telemetry: vscode_executeCommand {<br />  Metadata: {<br />    command: 'aws.amazonq.signout.commandPalette',<br />    duration: '412',<br />    result: 'Succeeded',<br />    awsAccount: 'not-set',<br />    awsRegion: 'not-set'<br />  },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:18:49.182 [debug] telemetry: auth_signInPageOpened {<br />  Metadata: { result: 'Succeeded', awsAccount: 'not-set', awsRegion: 'not-set' },<br />  Value: 1,<br />  Unit: 'None',<br />  Passive: true<br />}<br />2024-05-26 19:18:49.190 [info] toolkitApi: connection delete callback 06f56967-f6e1-47f0-9ff2-5e95a9b93d10<br />2024-05-26 19:18:49.275 [debug] Registered event handler for: onActiveConnectionModified<br />2024-05-26 19:18:49.275 [debug] Webview called command "refreshAuthState" with args: []<br />2024-05-26 19:18:49.283 [debug] Webview called command "getAuthState" with args: []<br />2024-05-26 19:18:49.286 [debug] Webview called command "getDefaultStartUrl" with args: []<br />2024-05-26 19:18:49.289 [debug] Webview called command "getRegions" with args: []<br />2024-05-26 19:18:49.289 [warning] AwsContext: no default region in credentials profile, falling back to us-east-1:<br />2024-05-26 19:18:49.290 [debug] Webview called command "fetchConnections" with args: []<br />2024-05-26 19:18:49.290 [debug] Webview called command "resetStoredMetricMetadata" with args: []<br />2024-05-26 19:18:49.668 [debug] Webview called command "refreshAuthState" with args: []<br />2024-05-26 19:18:49.669 [debug] Webview called command "getAuthState" with args: []<br />2024-05-26 19:19:02.741 [info] telemetry: sent batch (size=20)<br />2024-05-26 19:19:02.997 [info] telemetry: sent batch (size=15)<br /></pre> |

ivikash commented:
https://github.com/aws/aws-toolkit-vscode/assets/371007/22a4d71e-f026-4419-927c-af2e54b7302e

justinmk3 commented on packages/amazonq/package.json:
```diff
@@ -567,6 +588,36 @@
+                "command": "aws.codeWhisperer.toggleCodeScan.commandPalette",
+                "title": "%AWS.amazonq.toggleCodeScan%",
+                "category": "%AWS.amazonq.title%",
+                "enablement": "aws.codewhisperer.connected && aws.codewhisperer.connectionType != builderId"
```
there should be no need for a new `aws.codewhisperer.connectionType` global. a "toggle" command always looks the same in the palette, it's just named `Toggle Foo`. It shouldn't change its wording based on state, that is confusing, and also requires these extra global flags.

justinmk3 commented on packages/amazonq/package.json:
```diff
@@ -567,6 +588,36 @@
+                "enablement": "aws.codewhisperer.connected"
+            },
+            {
+                "command": "aws.amazonq.toggleCodeSuggestion.commandPalette",
```
As discussed, we can avoid these `.commandPalette` cases and re-use the existing commands.

Setting `source` may require some thought, but it's less important than avoiding lots of duplicate commands.

justinmk3 commented on packages/core/src/codecatalyst/commands.ts:
```diff
@@ -27,6 +27,7 @@ import { Auth } from '../auth/auth'
 import { SsoConnection } from '../auth/connection'
 import { isInDevEnv, isRemoteWorkspace } from '../shared/vscode/env'
 import { getShowManageConnections } from '../login/command'
+import { commandPalette } from '../codewhisperer/commands/types'
```
this "source" is pretty general so should probably live in somewhere higher up, e.g. constants.ts

justinmk3 commented on packages/core/src/codewhisperer/commands/basicCommands.ts:
```diff
@@ -53,16 +56,31 @@ export const toggleCodeSuggestions = Commands.declare(
             })
             vsCodeState.isFreeTierLimitReached = false
+
+            isSuggestionsEnabled
+                ? void showTimedMessage('Amazon Q: Auto-Suggestions are currently running.', MessageTimeOut)
+                : void showTimedMessage('Amazon Q: Auto-Suggestions are currently paused.', MessageTimeOut)
```
using a ternary may be clearer and less verbose

justinmk3 replied:
this kind of message is probably more appripriate as a status message rather than a popup

justinmk3 commented on packages/core/src/codewhisperer/commands/types.ts:
```diff
@@ -16,6 +16,8 @@ export const amazonQChatSource = 'amazonQChat'
 export const firstStartUpSource = ExtStartUpSources.firstStartUp
 /** Indicates a CodeWhisperer command was executed as a result of selecting an ellipses menu item */
 export const cwEllipsesMenu = 'ellipsesMenu'
+/** Indicates a CodeWhisperer command was executed from the command palette */
+export const commandPalette = 'commandPalette'
```
these should move up somewhere higher. they are not CW-specific

justinmk3 replied:
all of these constants should live in `packages/core/src/shared/telemetry/telemetry.ts` or perhaps a new `packages/core/src/shared/telemetry/constants.ts` file.

however, that can be done in a separate PR, to unblock this PR.

ivikash commented:
TODO:
- [ ] Update `Copy` for Error Messages for `Select Customization` and `Toggle Auto-Scans` when user is connected via AWS builder Id.
- [ ] Set proper `source` in `compositeKey` for Telemetry.

justinmk3 commented on packages/core/src/codecatalyst/commands.ts:
```diff
@@ -26,6 +26,7 @@ import { CreateDevEnvironmentRequest, UpdateDevEnvironmentRequest } from 'aws-sd
 import { Auth } from '../auth/auth'
 import { SsoConnection } from '../auth/connection'
 import { isInDevEnv, isRemoteWorkspace } from '../shared/vscode/env'
+import { commandPalette } from '../codewhisperer/commands/types'
```
the /codewhisperer/ dir is a hint that this constant should be moved somewhere higher. `packages/core/src/shared/telemetry/telemetry.ts` seems like a good place.

justinmk3 commented on packages/core/src/codewhisperer/commands/basicCommands.ts:
```diff
@@ -93,6 +107,12 @@ export const toggleCodeScans = Commands.declare(
             })
 
             await vscode.commands.executeCommand('aws.amazonq.refreshStatusBar')
+            void showTimedMessage(
```
```suggestion
            void vscode.window.setStatusBarMessage(
```

justinmk3 commented on packages/core/src/shared/vscode/commands2.ts:
```diff
@@ -520,7 +520,9 @@ function handleBadCompositeKey(data: { id: string; args: any[]; compositeKey: Co
         const indexAsInt = parseInt(index)
         const arg = args[indexAsInt]
-        if (field === 'source' && typeof arg !== 'string') {
+        if (field === 'source' && !arg) {
```
this would allow `arg=0` or `arg=''`, which is semantically not what we want. we specifically want to allow `undefined`, because that particular case is likely just "absence of an arg". whereas `0`, `''`, etc are not "absence" but rather "an arg was actually passed but likely a mistake".

justinmk3 replied:
```suggestion
        if (field === 'source' && arg === undefined) {
```
